src/main/java/hudson/plugins/ec2/util/MockAmazonEC2.java

```diff
@@ -0,0 +1,196 @@
+package hudson.plugins.ec2.util;
+
+import com.amazonaws.ClientConfiguration;
+import com.amazonaws.Request;
+import com.amazonaws.Response;
+import com.amazonaws.auth.AWSCredentialsProvider;
+import com.amazonaws.metrics.RequestMetricCollector;
+import com.amazonaws.services.ec2.AmazonEC2Client;
+import com.amazonaws.services.ec2.model.BlockDeviceMapping;
+import com.amazonaws.services.ec2.model.DescribeImagesRequest;
+import com.amazonaws.services.ec2.model.DescribeImagesResult;
+import com.amazonaws.services.ec2.model.DescribeInstancesRequest;
+import com.amazonaws.services.ec2.model.DescribeInstancesResult;
+import com.amazonaws.services.ec2.model.DescribeKeyPairsResult;
+import com.amazonaws.services.ec2.model.DescribeSecurityGroupsRequest;
+import com.amazonaws.services.ec2.model.DescribeSecurityGroupsResult;
+import com.amazonaws.services.ec2.model.DescribeSubnetsRequest;
+import com.amazonaws.services.ec2.model.DescribeSubnetsResult;
+import com.amazonaws.services.ec2.model.EbsBlockDevice;
+import com.amazonaws.services.ec2.model.Image;
+import com.amazonaws.services.ec2.model.Instance;
+import com.amazonaws.services.ec2.model.InstanceState;
+import com.amazonaws.services.ec2.model.InstanceStateChange;
+import com.amazonaws.services.ec2.model.InstanceStateName;
+import com.amazonaws.services.ec2.model.KeyPairInfo;
+import com.amazonaws.services.ec2.model.Reservation;
+import com.amazonaws.services.ec2.model.RunInstancesRequest;
+import com.amazonaws.services.ec2.model.RunInstancesResult;
+import com.amazonaws.services.ec2.model.SecurityGroup;
+import com.amazonaws.services.ec2.model.Subnet;
+import com.amazonaws.services.ec2.model.Tag;
+import com.amazonaws.services.ec2.model.TagSpecification;
+import com.amazonaws.services.ec2.model.TerminateInstancesRequest;
+import com.amazonaws.services.ec2.model.TerminateInstancesResult;
+import hudson.plugins.ec2.AmazonEC2Cloud;
+import hudson.plugins.ec2.EC2PrivateKey;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class MockAmazonEC2 extends AmazonEC2Client {

```

Comment:
Ok this definitely doesn't belong in the main source. To mock out a connection, you can use Mockito or PowerMockito to replace the connection instances being created at test time without adding test code at run time.

Reply:
@jvz well that's the issue really. How would you mock out a class that's loaded by an explicit `new` inside of a class loaded by the Jenkins plugin system in a running Jenkins instance?

Since the test is about testing that the functionality is properly triggered by Jenkins functionality I need to test it within a running Jenkins instance.

I *think* Jenkins instances have access to test classes when running with the JunitRunner, so I could move `MockAmazonEC2` to the `test` package, but then there'd be no compile-time knowledge of the class when choosing which implementation to load (i.e. real or test) so I'd have to use reflection which in my opinion is even dirtier.

So I'm really unsure how to test it in any other way.

Reply:
You might be able to use PowerMockito's static method mocking feature. Otherwise, the client factory should be refactored to be testable like you said.